pkg/webhook/preflight/nutanix/clients.go

```diff
@@ -83,126 +197,200 @@ func newClient(
 	page_ *int,
 	limit_ *int,
 	filter_ *string,
 	orderby_ *string,
 	apply_ *string,
 	select_ *string,
 	args ...map[string]interface{},
-) (*clustermgmtv4.ListClustersApiResponse, error) {
-	resp, err := c.v4client.ClustersApiInstance.ListClusters(
-		page_,
-		limit_,
-		filter_,
-		orderby_,
-		apply_,
-		select_,
-		args...,
-	)
-	if err != nil {
-		return nil, err
-	}
-	return resp, nil
+) (
+	*clustermgmtv4.ListClustersApiResponse,
+	error,
+) {
+	return callWithContext(ctx, func() (*clustermgmtv4.ListClustersApiResponse, error) {
+		return c.ListClustersFunc(
+			page_,
+			limit_,
+			filter_,
+			orderby_,
+			apply_,
+			select_,
+			args...,
+		)
+	})
 }
 
 func (c *clientWrapper) ListStorageContainers(
+	ctx context.Context,
 	page_ *int,
 	limit_ *int,
 	filter_ *string,
 	orderby_ *string,
 	select_ *string,
 	args ...map[string]interface{},
-) (*clustermgmtv4.ListStorageContainersApiResponse, error) {
-	resp, err := c.v4client.StorageContainerAPI.ListStorageContainers(
-		page_,
-		limit_,
-		filter_,
-		orderby_,
-		select_,
-		args...,
-	)
-	if err != nil {
-		return nil, err
-	}
-	return resp, nil
+) (
+	*clustermgmtv4.ListStorageContainersApiResponse,
+	error,
+) {
+	return callWithContext(ctx, func() (*clustermgmtv4.ListStorageContainersApiResponse, error) {
+		return c.ListStorageContainersFunc(
+			page_,
+			limit_,
+			filter_,
+			orderby_,
+			select_,
+			args...,
+		)
+	})
 }
 
-func (c *clientWrapper) GetSubnetById(id *string) (*netv4.GetSubnetApiResponse, error) {
-	resp, err := c.v4client.SubnetsApiInstance.GetSubnetById(id)
-	if err != nil {
-		return nil, err
-	}
-	return resp, nil
+func (c *clientWrapper) GetSubnetById(
+	ctx context.Context,
+	uuid *string,
+	args ...map[string]interface{},
+) (
+	*netv4.GetSubnetApiResponse,
+	error,
+) {
+	return callWithContext(ctx, func() (*netv4.GetSubnetApiResponse, error) {
+		return c.GetSubnetByIdFunc(
+			uuid,
+			args...,
+		)
+	})
 }
 
 func (c *clientWrapper) ListSubnets(
+	ctx context.Context,
 	page_ *int,
 	limit_ *int,
 	filter_ *string,
 	orderby_ *string,
 	expand_ *string,
 	select_ *string,
 	args ...map[string]interface{},
-) (*netv4.ListSubnetsApiResponse, error) {
-	resp, err := c.v4client.SubnetsApiInstance.ListSubnets(
-		page_,
-		limit_,
-		filter_,
-		orderby_,
-		expand_,
-		select_,
-		args...,
-	)
-	if err != nil {
-		return nil, err
+) (
+	*netv4.ListSubnetsApiResponse,
+	error,
+) {
+	return callWithContext(ctx, func() (*netv4.ListSubnetsApiResponse, error) {
+		return c.ListSubnetsFunc(
+			page_,
+			limit_,
+			filter_,
+			orderby_,
+			expand_,
+			select_,
+			args...,
+		)
+	})
+}
+
+// callWithContext is a helper function that immediately responds to context cancellation,
+// while calling a long-running, non-preemptible function. The long-running function always
+// runs to completion, but its result is only returned if the context is not cancelled.
+func callWithContext[T any](ctx context.Context, f func() (T, error)) (T, error) {
```

Comment:
👏

Reply:
Thanks! It took a bunch of iterations, but I found a way to do it! And then Sid made a couple suggestions that made it simpler.